common/service/config/ringpop.go

```diff
@@ -202,7 +210,65 @@ func (factory *RingpopFactory) getChannel(dispatcher *yarpc.Dispatcher) (*tcg.Ch
 	return ch, nil
 }
 
-func newDiscoveryProvider(cfg *Ringpop) (discovery.DiscoverProvider, error) {
+type dnsHostResolver interface {
+	LookupHost(ctx context.Context, host string) (addrs []string, err error)
+}
+
+type dnsProvider struct {
+	UnresolvedHosts []string
+	Resolver        dnsHostResolver
+	Logger          log.Logger
+}
+
+func newDNSProvider(hosts []string, resolver dnsHostResolver, logger log.Logger) *dnsProvider {
+	set := map[string]struct{}{}
+	for _, hostport := range hosts {
+		set[hostport] = struct{}{}
+	}
+
+	keys := []string{}
+	for key := range set {
+		keys = append(keys, key)
+	}
+	return &dnsProvider{
+		UnresolvedHosts: keys,
+		Resolver:        resolver,
+		Logger:          logger,
+	}
+}
+
+func (provider *dnsProvider) Hosts() ([]string, error) {
+	results := []string{}
+	resolvedHosts := map[string][]string{}
+	for _, hostport := range provider.UnresolvedHosts {
+		host, port, err := net.SplitHostPort(hostport)
+		if err != nil {
+			provider.Logger.Warn("Could not split host and port", tag.Address(hostport), tag.Error(err))
+			continue
+		}
+
+		resolved, exists := resolvedHosts[host]
+		if !exists {
+			resolved, err = provider.Resolver.LookupHost(context.Background(), host)
+			if err != nil {
+				provider.Logger.Warn("Could not resolve host", tag.Address(host), tag.Error(err))
+				continue
+			}
+			for _, r := range resolved {
```

Comment:
are there reasons to do a slice copy here instead of simply doing  ```resolvedHosts[host] = resolved``` ?

Reply:
🤦‍♂ Nope no reason. I updated it to be direct assignment.